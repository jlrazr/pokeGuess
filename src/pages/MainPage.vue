<template>
  <div class="container">
    <h1 v-if="!pickedPokemon">Getting pokemons out of their pokeballs...</h1>
    <div class="container" v-else>
      <h2>Can you guess the Pok√©mon?</h2>
      <PokemonPic :pokeId="pickedPokemon.data.id" :reveal="revealPokemon" />
      <PokemonOptions :pokemons="pokemonsArray" @pick-pokemon="checkAnswer" />
      <h2 class="fade-in message">
        {{ message }}
      </h2>
      <button class="fade-in" v-if="playerLose || playerWin" @click="newGame">
        Play again!
      </button>
    </div>
    <h3>Score: {{ score }}</h3>
    <h3>Personal record: {{ record }}</h3>
  </div>
</template>

<script>
import PokemonOptions from "@/components/PokemonOptions.vue";
import PokemonPic from "@/components/PokemonPic.vue";
import { getPokemonOptions } from "@/utils/getPokemonOptions.js";

export default {
  name: "MainPage",
  components: {
    PokemonOptions,
    PokemonPic,
  },

  data() {
    return {
      pokemonsArray: [],
      pickedPokemon: null,
      revealPokemon: false,
      playerWin: false,
      playerLose: false,
      message: "",
      score: 0,
      record: localStorage.getItem("record") || 0,
    };
  },

  methods: {
    async mixPokemonArray() {
      const randomPick = Math.floor(Math.random() * 4);
      this.pokemonsArray = await getPokemonOptions();
      this.pickedPokemon = this.pokemonsArray[randomPick];
    },
    checkAnswer(pokemonId) {
      this.revealPokemon = true;
      if (pokemonId === this.pickedPokemon.data.id) {
        this.playerWin = true;
        this.score++;
        this.message = `Yes! The pokemon is ${this.pickedPokemon.data.name}`;
        this.setRecord();
      } else {
        this.playerLose = true;
        this.score = 0;
        this.message = `Sorry, the hidden pokemon was ${this.pickedPokemon.data.name}`;
      }
    },
    newGame() {
      this.revealPokemon = false;
      this.pickedPokemon = null;
      this.message = "";
      this.playerWin = false;
      this.playerLose = false;
      this.pokemonsArray = [];
      this.record = localStorage.getItem("record");
      this.mixPokemonArray();
    },
    setRecord() {
      const currScore = this.score;
      const currRecord = localStorage.getItem("record");

      if (currScore > currRecord) {
        localStorage.setItem("record", currScore);
      }
    },
  },

  beforeMount() {
    this.mixPokemonArray();
  },
};
</script>
<style lang="scss" scoped>
.container {
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  z-index: 1;

  h2 {
    color: #fff;
    font-size: 22px;

    @media (min-width: 766px) {
      font-size: 32px;
    }
  }

  h3 {
    color: #fff;
    font-size: 16px;
    margin: 0;

    @media (min-width: 766px) {
      font-size: 32px;
    }
  }

  button {
    margin-right: 10px;
    background: #fff;
    border-radius: 5px;
    border: 1px solid rgba(0, 0, 0, 0.2);
    cursor: pointer;
    margin-bottom: 10px;
    width: 250px;
    padding: 14px 5px;
    font-weight: 600;
  }

  span {
    color: green;
    font-weight: 700;
  }

  .message {
    font-size: 14px;

    @media (min-width: 766px) {
      font-size: 32px;
    }
  }
}
</style>
