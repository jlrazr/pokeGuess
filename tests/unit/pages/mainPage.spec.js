import { shallowMount, mount } from "@vue/test-utils";
import MainPage from "@/pages/MainPage";
import { pokemons } from "../mocks/pokemons.mock";

describe("MainPage Component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(MainPage);
  });

  test("must match the snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("must execute mixPokemonArray function on beforeMount", () => {
    const mixPokemonSpy = jest.spyOn(MainPage.methods, "mixPokemonArray");
    const wrapper = shallowMount(MainPage);
    expect(mixPokemonSpy).toHaveBeenCalled();
  });

  test("must match with snapshot after pokemons have been loaded", () => {
    const wrapper = shallowMount(MainPage, {
      data() {
        return {
          pokemonsArray: pokemons,
          pickedPokemon: pokemons[0],
          revealPokemon: false,
          playerWin: false,
          playerLose: false,
          message: "",
          score: 0,
          record: 0,
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test("must show the PokemonPic and PokemonOptions components mounted and redered", () => {
    const wrapper = shallowMount(MainPage, {
      data() {
        return {
          pokemonsArray: pokemons,
          pickedPokemon: pokemons[0],
          revealPokemon: false,
          playerWin: false,
          playerLose: false,
          message: "",
          score: 0,
          record: 0,
        };
      },
    });

    const pokePic = wrapper.find("pokemon-pic-stub");
    const options = wrapper.find("pokemon-options-stub");

    expect(pokePic.exists()).toBeTruthy();
    expect(pokePic.attributes("pokeid")).toBe("150");

    expect(options.exists()).toBeTruthy();
    expect(options.attributes("pokemons")).toBeTruthy();
  });

  test("tests the checkAnswer function", async () => {
    const wrapper = shallowMount(MainPage, {
      data() {
        return {
          pokemonsArray: pokemons,
          pickedPokemon: pokemons[0],
          revealPokemon: false,
          playerWin: false,
          playerLose: false,
          message: "",
          score: 0,
          record: 0,
        };
      },
    });

    await wrapper.vm.checkAnswer(150);

    expect(wrapper.find("button"));
    expect(wrapper.vm.revealPokemon).toBe(true);
    expect(wrapper.vm.message).toBe(
      `Yes! The pokemon is ${pokemons[0].data.name}`
    );

    await wrapper.vm.checkAnswer(111);
    expect(wrapper.vm.message).toBe(
      `Sorry, the hidden pokemon was ${pokemons[0].data.name}`
    );
  });
});
