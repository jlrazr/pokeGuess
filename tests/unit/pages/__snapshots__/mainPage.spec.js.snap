// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage Component must match the snapshot 1`] = `
<div class="container">
  <h1>Getting pokemons out of their pokeballs...</h1>
  <h3>Score: 0</h3>
  <h3>Personal record: 0</h3>
</div>
`;

exports[`MainPage Component must match with snapshot after pokemons have been loaded 1`] = `
<div class="container">
  <div class="container">
    <h2>Can you guess the Pok√©mon?</h2>
    <pokemon-pic-stub pokeid="150" reveal="false"></pokemon-pic-stub>
    <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
    <h2 class="fade-in message"></h2>
    <!--v-if-->
  </div>
  <h3>Score: 0</h3>
  <h3>Personal record: 0</h3>
</div>
`;
